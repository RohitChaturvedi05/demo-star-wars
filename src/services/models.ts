import { CharacterDetails, Film } from '../models/Character';

export type CharacterResponse = {
    properties: {
        created: string;
        edited: string;
        name: string;
        gender: string;
        skin_color: string;
        hair_color: string;
        height: string;
        eye_color: string;
        mass: string;
        homeworld: string;
        birth_year: string;
        url: string;
    };
    _id: string;
    description: string;
    uid: string;
};

export type StarShipResponse = {
    name: string;
    cargo_capacity: string;
    passengers: string;
    max_atmosphering_speed: string;
    manufacturer: string;
};

export type PlanetResponse = {
    result: {
        uid: string;
        properties: {
            name: string;
            diameter: string;
            gravity: string;
            population: string;
        };
    };
};

export type CharacterByIdResponse = Omit<CharacterDetails, 'id'>;

export type FilmResponse = Film;
