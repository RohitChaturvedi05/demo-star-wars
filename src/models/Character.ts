export type Planet = {
    id: string;
    name: string;
    diameter: string;
    gravity: string;
    population: string;
};

export type Film = {
    id: string;
    title: string;
    director: string;
    producer: string;
    release_date: string;
};

export type Starship = {
    name: string;
    cargo_capacity: string;
    passengers: string;
    max_atmosphering_speed: string;
    manufacturer: string;
};

export type Character = {
    uid: string;
    id: string;
    name: string;
    gender: string;
    skin_color: string;
    hair_color: string;
    height: string;
    eye_color: string;
    mass: string;
    birth_year: string;
    homeworldUrl: string;
    homeworld: string;
};

export type CharacterDetails = Character & {
    films: string[];
    starships: string[];
};

export type Characters = Character[];
