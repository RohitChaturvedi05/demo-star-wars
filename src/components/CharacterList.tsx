import ListAltIcon from '@mui/icons-material/ListAlt';
import StarFilled from '@mui/icons-material/Star';
import StarBorderOutlined from '@mui/icons-material/StarBorderOutlined';
import {
    Box,
    IconButton,
    Pagination,
    Paper,
    Table,
    TableBody,
    TableCell,
    TableContainer,
    TableHead,
    TableRow,
} from '@mui/material';

import { useMemo } from 'react';
import { Characters, Planet } from '../models/Character';
import { State } from '../state/reducer';
import { Loader } from './Loader';

export type CharacterListProps = {
    characters: Characters;
    planets: Record<string, Planet>;
    favCharacters: State['favCharacters'];
    isLoading?: boolean;
    showPagination?: boolean;
    pagination?: {
        currentPage: number;
        totalPage: number;
    };
    onFavoriteToggle?: (id: string) => void;
    onDetailClick?: (id: string) => void;
    onPageChange?: (page: number) => void;
};

export const CharacterList = ({
    characters,
    favCharacters,
    pagination,
    showPagination = false,
    isLoading = false,
    planets,
    onFavoriteToggle,
    onDetailClick,
    onPageChange = () => {},
}: CharacterListProps) => {
    const shouldShowFavorite = !!onFavoriteToggle;
    const shouldShowDetails = !!onDetailClick;

    const displayCharacters = useMemo(() => {
        return characters.map(({ uid, name, gender, homeworld }) => {
            const isFav = favCharacters.includes(uid);
            return (
                <TableRow key={uid}>
                    <TableCell>
                        {shouldShowFavorite && (
                            <IconButton
                                data-testid={`btn-fav-${uid}`}
                                onClick={() => onFavoriteToggle(uid)}
                            >
                                {isFav && <StarFilled />}
                                {!isFav && <StarBorderOutlined />}
                            </IconButton>
                        )}
                    </TableCell>
                    <TableCell>{name}</TableCell>
                    <TableCell>{gender}</TableCell>
                    <TableCell>{planets[homeworld]?.name ?? ''}</TableCell>
                    {shouldShowDetails && (
                        <TableCell>
                            <IconButton
                                data-testid={`btn-details-${uid}`}
                                onClick={() => onDetailClick(uid)}
                            >
                                <ListAltIcon />
                            </IconButton>
                        </TableCell>
                    )}
                </TableRow>
            );
        });
    }, [
        characters,
        favCharacters,
        shouldShowFavorite,
        planets,
        shouldShowDetails,
        onFavoriteToggle,
        onDetailClick,
    ]);

    return (
        <TableContainer component={Paper}>
            <Table>
                <TableHead>
                    <TableRow>
                        {shouldShowFavorite && <TableCell></TableCell>}
                        <TableCell>Name</TableCell>
                        <TableCell>Gender</TableCell>
                        <TableCell>Home planet</TableCell>
                        {shouldShowDetails && <TableCell></TableCell>}
                    </TableRow>
                </TableHead>
                <TableBody>
                    {characters.length > 0 && !isLoading && displayCharacters}
                    {isLoading && (
                        <TableRow>
                            <TableCell colSpan={4}>
                                <Loader />
                            </TableCell>
                        </TableRow>
                    )}
                    {characters.length === 0 && !isLoading && (
                        <TableRow>
                            <TableCell colSpan={4}>No Records Found</TableCell>
                        </TableRow>
                    )}
                </TableBody>
            </Table>
            {showPagination && !isLoading && (
                <Box
                    sx={{
                        display: 'flex',
                        justifyContent: 'flex-end',
                        p: 2,
                    }}
                >
                    <Pagination
                        data-testid="character-list$pagination"
                        disabled={pagination?.totalPage === 0}
                        page={pagination?.currentPage ?? 0}
                        count={pagination?.totalPage ?? 0}
                        shape="rounded"
                        onChange={(_, page) => onPageChange(page)}
                    />
                </Box>
            )}
        </TableContainer>
    );
};
