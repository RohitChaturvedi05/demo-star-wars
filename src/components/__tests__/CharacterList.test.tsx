import { render, screen } from '@testing-library/react';
import { Characters } from '../../models/Character';
import { CharacterList, CharacterListProps } from '../CharacterList';

const mockProps: CharacterListProps = {
    characters: [],
    planets: {},
    favCharacters: [],
    showPagination: false,
    pagination: {
        currentPage: 1,
        totalPage: 1,
    },
    onFavoriteToggle: jest.fn(),
    onDetailClick: jest.fn(),
    onPageChange: jest.fn(),
};
const mockCharacters: Characters = [
    {
        birth_year: '19BBY',
        eye_color: 'blue',
        gender: 'male',
        hair_color: 'blond',
        height: '172',
        homeworld: '1',
        homeworldUrl: 'https://www.swapi.tech/api/planets/1',
        id: '5f63a36eee9fd7000499be42',
        mass: '77',
        name: 'Luke Skywalker',
        skin_color: 'fair',
        uid: '1',
    },
    {
        birth_year: '112BBY',
        eye_color: 'blue',
        gender: 'male',
        hair_color: 'blond',
        height: '172',
        homeworld: '2',
        homeworldUrl: 'https://www.swapi.tech/api/planets/2',
        id: 'asdsad21122',
        mass: '77',
        name: 'Darth Vader',
        skin_color: 'fair',
        uid: '2',
    },
];
describe('CharacterList', () => {
    it('should render without crashing', () => {
        const { container } = render(<CharacterList {...mockProps} />);
        expect(container).toMatchSnapshot();
    });

    it('should render list of characters when data is loaded', () => {
        render(<CharacterList {...mockProps} characters={mockCharacters} />);

        expect(screen.getByText('Luke Skywalker')).toBeInTheDocument();
        expect(screen.getByText('Darth Vader')).toBeInTheDocument();
    });

    it('should handle empty characters list', () => {
        render(<CharacterList {...mockProps} characters={[]} />);
        expect(screen.getByText(/No Records Found/i)).toBeInTheDocument();
    });
    it('should have pagination component when showPagination is true', () => {
        render(
            <CharacterList
                {...mockProps}
                characters={mockCharacters}
                showPagination
            />
        );
        expect(
            screen.getByTestId('character-list$pagination')
        ).toBeInTheDocument();
    });
});
