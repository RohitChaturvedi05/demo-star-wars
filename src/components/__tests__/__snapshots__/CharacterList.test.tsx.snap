// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList should render without crashing 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <table
      class="MuiTable-root css-1xwxv7r-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            Gender
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            Home planet
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          />
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            colspan="4"
          >
            No Records Found
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
